### ISSUE TITLE
JarvisCFID numpy array handling error

### ISSUE THREAD

**Issue by lensonp** [hackingmaterials/matminer#785]
I tried to submit to the discourse forum, but the link given here appears to be broken.

matminer==0.7.6
pymatgen==2022.1.24
numpy==1.22.1

Running JarvisCFID for a bunch of materials, I am seeing a fair number of crashes due to these:
```
venv/lib/python3.8/site-packages/numpy/core/_methods.py:163: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
```

It does crash the featurization, and the resulting stack trace is something like this:
```
  File "run_jarvis.py", line 42, in run_one_jarvis
    feats = j.featurize(struct)
  File "venv/lib/python3.8/site-packages/matminer/featurizers/structure/composite.py", line 110, in featurize
    mean_chem = np.mean(arr, axis=0)
  File "<__array_function__ internals>", line 180, in mean
  File "venv/lib/python3.8/site-packages/numpy/core/fromnumeric.py", line 3474, in mean
    return _methods._mean(a, axis=axis, dtype=dtype,
  File "venv/lib/python3.8/site-packages/numpy/core/_methods.py", line 179, in _mean
    ret = umr_sum(arr, axis, dtype, out, keepdims, where=where)
ValueError: operands could not be broadcast together with shapes (438,) (0,) 
```
... finally, here is an example pymatgen structure that throws the error, as JSON. 
```
{"@module": "pymatgen.core.structure", "@class": "Structure", "charge": null, "lattice": {"matrix": [[4.285388, 0.0, 0.0], [0.0, 7.277622, 0.0], [0.0, 0.0, 8.124477]], "a": 4.285388, "b": 7.277622, "c": 8.124477, "alpha": 90.0, "beta": 90.0, "gamma": 90.0, "volume": 253.38159011912964}, "sites": [{"species": [{"element": "Li", "occu": 1}], "abc": [0.25, 0.639488, 0.559321], "xyz": [1.071347, 4.653951937535999, 4.544190600117], "label": "Li", "properties": {"magmom": -0.0}}, {"species": [{"element": "Li", "occu": 1}], "abc": [0.75, 0.360512, 0.440679], "xyz": [3.214041, 2.623670062464, 3.580286399883], "label": "Li", "properties": {"magmom": 0.005}}, {"species": [{"element": "Li", "occu": 1}], "abc": [0.75, 0.860512, 0.059321], "xyz": [3.214041, 6.262481062464, 0.481952100117], "label": "Li", "properties": {"magmom": 0.008}}, {"species": [{"element": "Li", "occu": 1}], "abc": [0.25, 0.139488, 0.940679], "xyz": [1.071347, 1.015140937536, 7.642524899883001], "label": "Li", "properties": {"magmom": 0.001}}, {"species": [{"element": "Eu", "occu": 1}], "abc": [0.25, 0.524009, 0.189861], "xyz": [1.071347, 3.8135394265979996, 1.5425213276970002], "label": "Eu", "properties": {"magmom": 6.828}}, {"species": [{"element": "Eu", "occu": 1}], "abc": [0.75, 0.475991, 0.810139], "xyz": [3.214041, 3.464082573402, 6.581955672303001], "label": "Eu", "properties": {"magmom": -6.824}}, {"species": [{"element": "Eu", "occu": 1}], "abc": [0.75, 0.975991, 0.689861], "xyz": [3.214041, 7.1028935734020004, 5.604759827697], "label": "Eu", "properties": {"magmom": 6.834}}, {"species": [{"element": "Eu", "occu": 1}], "abc": [0.25, 0.024009, 0.310139], "xyz": [1.071347, 0.174728426598, 2.5197171723030003], "label": "Eu", "properties": {"magmom": 6.852}}, {"species": [{"element": "P", "occu": 1}], "abc": [0.25, 0.77183, 0.891867], "xyz": [1.071347, 5.6170869882600005, 7.245952928559], "label": "P", "properties": {"magmom": -0.022}}, {"species": [{"element": "P", "occu": 1}], "abc": [0.75, 0.22817, 0.108133], "xyz": [3.214041, 1.6605350117400002, 0.8785240714410001], "label": "P", "properties": {"magmom": -0.037}}, {"species": [{"element": "P", "occu": 1}], "abc": [0.75, 0.72817, 0.391867], "xyz": [3.214041, 5.29934601174, 3.1837144285590004], "label": "P", "properties": {"magmom": -0.058}}, {"species": [{"element": "P", "occu": 1}], "abc": [0.25, 0.27183, 0.608133], "xyz": [1.071347, 1.97827598826, 4.940762571441001], "label": "P", "properties": {"magmom": -0.021}}]}
```

To load this into a pmg structure and plug it into Jarvis, copy/paste this JSON to a file and then do this:
```
from pymatgen.core.structure import Structure
from matminer.featurizers.structure import JarvisCFID
struct = Structure.from_dict(json.load(open(structfile,'r')))
feats = j.featurize(struct)
```

**Comment by ardunn** [comment]
Hi @lensonp 

I've fixed the link on the readme, thanks for pointing it out!


Could you provide a structure file that this happens for? Or does it happen for all the structures you try?

**Comment by lensonp** [comment]
Well, I couldn't attach a .json file, so I gave the text of the .json above. You would have to save it to a .json file to test it by the script I posted there. That particular json represents Materials Project record mp-1211143.

I'm running this featurizer on a large collection of materials- the failures aren't an issue for me but I thought you'd like to know. From the error message, it seems like you might be able to fix it by declaring a dtype for an array in certain cases.

**Comment by ardunn** [comment]
@lensonp I realized these errors happen because the source jarvis chemical descriptor files actually do not have some elements (including Eu). The source code I ported over from the original Jarvis publication simply assigned missing elements an empty array of descriptors instead of throwing an informative error. The code now throws a more informative error instead of assigning an empty array; this is a safer option than just ignoring it and taking the mean of the remaining elemental descriptors, since that would cause changes in the descriptors with no warnings to bring to the user's attention.
